### PR TITLE
[WIP] Switching tests to pytest and removing nose dependencies

### LINKED ISSUE #1280: Switch testing to pytest?
I have a feeling we've discussed this before. Nose is no longer being developed, and from what I have heard and experienced myself, it looks like pytest (https://docs.pytest.org/en/latest/) is a good way to go. Looks like maybe nose2 is also an option (http://nose2.readthedocs.io/en/latest/).

### PR DESCRIPTION
**Pull request by thechargedneutron**
Fixes #1280 

Please review if this is what you intended. I have added separate check function and used it.

### CODE REVIEW

**Comment by thechargedneutron** [comment]
@skoudoro  Can you please have a look at this PR and suggest what changes should be done?

File: .cache/v/cache/lastfailed [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "dipy/workflows/tests/test_docstring_parser.py": true, 
+  "dipy/workflows/tests/test_iap.py": true, 
+  "dipy/workflows/tests/test_masking.py": true, 
+  "dipy/workflows/tests/test_reconst_csa_csd.py": true, 
+  "dipy/workflows/tests/test_reconst_dti.py": true, 
+  "dipy/workflows/tests/test_segment.py": true, 
+  "dipy/workflows/tests/test_workflow.py": true
+}
```
Comment (skoudoro) [inline .cache/v/cache/lastfailed]: This file should be removed

File: .travis.yml [hunk 2]
```diff
@@ -86,7 +86,8 @@ before_install:
     - python --version # just to check
     # Needed for Python 3.5 wheel fetching
     - $PIPI --upgrade pip setuptools
-    - $PIPI nose;
+    - $PIPI nose
```
Comment (skoudoro) [inline .travis.yml]: we do not want `nose` anymore so you can remove this line

File: dipy/reconst/benchmarks/bench_bounding_box.py [hunk 3]
```diff
@@ -7,13 +7,13 @@
     import dipy.reconst as dire
     dire.bench()
 
-If you have doctests enabled by default in nose (with a noserc file or
+If you have doctests enabled by default in ashu (with a ashurc file or
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_bounding_box.py]: what is `ashu` ?

File: dipy/reconst/benchmarks/bench_bounding_box.py [hunk 4]
```diff
@@ -7,13 +7,13 @@
...
 environment variable), and you have a numpy version <= 1.6.1, this will also
 run the doctests, let's hope they pass.
 
 Run this benchmark with:
 
-    nosetests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_bounding_box.py
+    ashutests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_bounding_box.py
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_bounding_box.py]: what is `ashutests` ?

File: dipy/reconst/benchmarks/bench_peaks.py [hunk 5]
```diff
@@ -5,13 +5,13 @@
     import dipy.reconst as dire
     dire.bench()
 
-If you have doctests enabled by default in nose (with a noserc file or
+If you have doctests enabled by default in ashu (with a ashurc file or
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_peaks.py]: same comment as below

File: dipy/reconst/benchmarks/bench_peaks.py [hunk 6]
```diff
@@ -5,13 +5,13 @@
...
 environment variable), and you have a numpy version <= 1.6.1, this will also
 run the doctests, let's hope they pass.
 
 Run this benchmark with:
 
-    nosetests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_peaks.py
+    ashutests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_peaks.py
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_peaks.py]: same comment as below

File: dipy/reconst/benchmarks/bench_squash.py [hunk 7]
```diff
@@ -5,14 +5,14 @@
     import dipy.reconst as dire
     dire.bench()
 
-If you have doctests enabled by default in nose (with a noserc file or
+If you have doctests enabled by default in ashu (with a ashurc file or
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_squash.py]: same comment as above

File: dipy/reconst/benchmarks/bench_squash.py [hunk 8]
```diff
@@ -5,14 +5,14 @@
...
 environment variable), and you have a numpy version <= 1.6.1, this will also
 run the doctests, let's hope they pass.
 
 
 Run this benchmark with:
 
-    nosetests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_squash.py
+    ashutests -s --match '(?:^|[\\b_\\.//-])[Bb]ench' /path/to/bench_squash.py
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_squash.py]: same comment as above

File: dipy/reconst/benchmarks/bench_squash.py [hunk 9]
```diff
@@ -121,7 +121,7 @@ def old_squash(arr, mask=None, fill=0):
 
 
 def bench_quick_squash():
-    # nosetests -s --match '(?:^|[\\b_\\.//-])[Bb]ench'
+    # ashutests -s --match '(?:^|[\\b_\\.//-])[Bb]ench'
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_squash.py]: same comment as above

File: dipy/reconst/benchmarks/bench_vec_val_sum.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
     import dipy.reconst as dire
     dire.bench()
 
-If you have doctests enabled by default in nose (with a noserc file or
+If you have doctests enabled by default in ashu (with a ashurc file or
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_vec_val_sum.py]: same comment as above

File: dipy/reconst/benchmarks/bench_vec_val_sum.py [hunk 11]
```diff
@@ -26,7 +26,7 @@ def with_einsum(f): return f
 
 @with_einsum
 def bench_vec_val_vect():
-    # nosetests -s --match '(?:^|[\\b_\\.//-])[Bb]ench'
+    # ashutests -s --match '(?:^|[\\b_\\.//-])[Bb]ench'
```
Comment (skoudoro) [inline dipy/reconst/benchmarks/bench_vec_val_sum.py]: same comment as above

File: dipy/reconst/recspeed.html [hunk 12]
```diff
@@ -0,0 +1,4329 @@
+<!DOCTYPE html>
```
Comment (skoudoro) [inline dipy/reconst/recspeed.html]: file to remove

**Comment by thechargedneutron** [comment]
@skoudoro  Sorry for the sloppy work in this PR. `ashu` was only used to mask the occurrences of the word `nose`. Really sorry for this.
Also the `.cache` and `.html` files are removed. But I don't why it was created.

**Comment by thechargedneutron** [comment]
@skoudoro  Can you provide an example where we are going to use `pytest coverage` and `pytest bechmark`? I looked into those things but could not make out where to use.

**Comment by skoudoro** [comment]
Hi @thechargedneutron, it seems we change directory before running the tests. This is why this line does not work. we need to look how can we run `pytest` from package installation and not from the source.

File: .travis.yml [hunk 13]
```diff
@@ -127,6 +127,6 @@ script:
       cp ../.coveragerc .;
       COVER_ARGS="--with-coverage --cover-package dipy";
       fi
-    - nosetests --with-doctest --verbose $COVER_ARGS dipy
+    - pytest --doctest-modules --durations=10 -svv $COVER_ARGS dipy
```
Comment (skoudoro) [inline .travis.yml]: So, following [here](https://docs.pytest.org/en/latest/usage.html#cmdline), you just need to add `--pyargs` option. 
Here, the new line : 

` pytest --doctest-modules --durations=10 -svv --pyargs dipy`

**Comment by skoudoro** [comment]
Great! there is some progress. It seems we need to avoid `numpy.testing` because it tries to import `nose`. This means you have to replace all `numpy.testing` functions.

- It seems that `pytest` manage correctly `Numpy` array so you can use what you created before (e.g `assert_equal`, ...)
- you need to change `assert_almost_equal` by using `pytest.approx` which is compatible with numpy array and all basic data types.
- [here](https://github.com/pytest-dev/pytest/pull/2492/files#diff-a4d4d6cc8ecc32438c8a4da391852f22R325), you can see how to do it with `pytest`. 

First, try it locally with 1 file like `dipy.direction.tests.test_peaks.py` before doing the change in all the package.

Let me know if you have any big trouble. Thanks again for this works!

File: dipy/core/tests/test_optimize.py [hunk 14]
```diff
@@ -57,7 +58,7 @@ def test_optimize_new_scipy():
     npt.assert_array_almost_equal(opt.xopt, np.array([0, 0, 0, 0.]))
 
 
-@npt.dec.skipif(not SCIPY_LESS_0_12)
+@pytest.mark.skipif(not SCIPY_LESS_0_12)
```
Comment (skoudoro) [inline dipy/core/tests/test_optimize.py]: As you can see [here](https://travis-ci.org/nipy/dipy/jobs/355024064#L5253),  `pytest` need the parameter `reason`. so you need to update it like this : 
`@pytest.mark.skipif(not SCIPY_LESS_0_12, reason="requires scipy < 0.12")`

File: dipy/utils/testing.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+def assert_almost_equal(actual, desired, decimal=7):
+	if not abs(desired - actual) < 1.5 * 10**(-decimal):
+		raise AssertionError("Values are not almost equal to {} decimals".format(decimal))
+
+def assert_array_equal(actual, desired):
+	assert np.all(actual == approx(desired))
+
+def assert_array_almost_equal(actual, desired, decimal=7):
+	assert np.all(actual == approx(desired, rel=10**(-1*decimal)))
```
Comment (skoudoro) [inline dipy/utils/testing.py]: 1. Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem
2. I think `np.all` is not necessary. you can just do  `assert actual == approx(desired`, rel=10**(-1*decimal))

File: dipy/utils/testing.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+def assert_(statement):
+	assert statement
```
Comment (skoudoro) [inline dipy/utils/testing.py]: Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem

File: dipy/utils/testing.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+def assert_array_less(first, second):
+	assert np.all(first < second)
```
Comment (skoudoro) [inline dipy/utils/testing.py]: Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem

File: dipy/utils/testing.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+
+'''Credits to the below functions goes to numpy.testing package'''
+
+
+def assert_allclose(actual, desired, rtol=1e-7, atol=0, equal_nan=True, err_msg='', verbose=True):
+	def compare(x, y):
+		return np.core.numeric.isclose(x, y, rtol=rtol, atol=atol, equal_nan=equal_nan)
+
+	actual, desired = np.asanyarray(actual), np.asanyarray(desired)
```
Comment (skoudoro) [inline dipy/utils/testing.py]: Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem

File: dipy/utils/testing.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+		if hasattr(excClass, '__name__'): excName = excClass.__name__
+		else: excName = str(excClass)
+		raise AssertionError("{} not raised by {}".format(excName, callableObj))
+
+def assert_almost_equal(actual, desired, decimal=7):
+	if not abs(desired - actual) < 1.5 * 10**(-decimal):
+		raise AssertionError("Values are not almost equal to {} decimals".format(decimal))
+
+def assert_array_equal(actual, desired):
+	assert np.all(actual == approx(desired))
```
Comment (skoudoro) [inline dipy/utils/testing.py]: 1. Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem
2. I think `np.all` is not necessary. you can just do  `assert actual == desired`

File: dipy/utils/testing.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+	except excClass:
+		return
+	else:
+		if hasattr(excClass, '__name__'): excName = excClass.__name__
+		else: excName = str(excClass)
+		raise AssertionError("{} not raised by {}".format(excName, callableObj))
+
+def assert_almost_equal(actual, desired, decimal=7):
+	if not abs(desired - actual) < 1.5 * 10**(-decimal):
+		raise AssertionError("Values are not almost equal to {} decimals".format(decimal))
```
Comment (skoudoro) [inline dipy/utils/testing.py]: 1. Like you did before, can you add  `try..... except AssertionError.....` for getting more information when there is a problem

**Comment by skoudoro** [comment]
`[31m[1m= 118 failed, 598 passed, 18 skipped, 63 warnings, 43 error in 1109.22 seconds`

80% of tests passed -> you are close :-)

File: dipy/utils/testing.py (hunk 15 above)
Comment (skoudoro) [reply dipy/utils/testing.py]: Hi @thechargedneutron, I think you are close, Any chance to make this change?

File: dipy/utils/testing.py [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+from numpy.core.numerictypes import issubdtype
+from numpy.core.fromnumeric import any as npany
+import numpy as np
+
+
+def assert_equal(value1, value2):
+	if not 'numpy' in str(type(value1)):
+		value1 = np.array(value1)
+	if not 'numpy' in str(type(value2)):
+		value2 = np.array(value2)
```
Comment (skoudoro) [inline dipy/utils/testing.py]: Why are you doing this test?

File: dipy/utils/testing.py [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+	try:
+		assert np.all(value1 == value2)
```
Comment (skoudoro) [inline dipy/utils/testing.py]: why are you using `np.all`? Not needed I think

File: dipy/utils/testing.py [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+	except AssertionError:
+		raise AssertionError(str(value1) + ' != ' + str(value2))
+
+def assert_not_equal(value1, value2):
+	if not 'numpy' in str(type(value1)):
+		value1 = np.array(value1)
+	if not 'numpy' in str(type(value2)):
+		value2 = np.array(value2)
```
Comment (skoudoro) [inline dipy/utils/testing.py]: same comment as below

File: dipy/utils/testing.py (hunk 22 above)
Comment (skoudoro) [reply dipy/utils/testing.py]: as you can see [here](https://github.com/pytest-dev/pytest/pull/2492/files#diff-a4d4d6cc8ecc32438c8a4da391852f22R325),  `pytest` manage numpy array and list of numpy array, so I still think that `np.all` is not needed

File: dipy/utils/testing.py (hunk 21 above)
Comment (skoudoro) [reply dipy/utils/testing.py]: as you can see [here](https://github.com/pytest-dev/pytest/pull/2492/files#diff-a4d4d6cc8ecc32438c8a4da391852f22R325),  `pytest` manage numpy array and list of numpy array, so I am really not sure we need this test

**Comment by skoudoro** [comment]
superseded by #1652. Closing

**Pull request closed without merging**